src/main/scala/io/iohk/ethereum/mpt/MerklePatriciaTrie.scala

```diff
@@ -42,131 +46,44 @@ object MerklePatriciaTrie {
-
-  private def matchingLength(a: Array[Byte], b: Array[Byte]): Int = {
-    var prefixLen = 0
-    while (prefixLen < a.length && prefixLen < b.length && a(prefixLen) == b(prefixLen)) {
-      prefixLen = prefixLen + 1
+    if (EmptyRootHash sameElements rootHash)
+      MerklePatriciaTrie(source)
+    else {
+      val root = getFromHash(rootHash, source)
+      new MerklePatriciaTrie[K, V](Some(root), source)(kSerializer, vSerializer)
     }
-    prefixLen
   }
 
-  private def updateNodesInStorage(
-    previousRootHash: Array[Byte],
-    newRoot: Option[MptNode],
-    toRemove: Seq[MptNode],
-    toUpdate: Seq[MptNode],
-    nodeStorage: NodesKeyValueStorage): NodesKeyValueStorage = {
-    val rootCapped = newRoot.map(_.capped).getOrElse(ByteString.empty)
-    val toBeRemoved = toRemove.filter { node =>
-      val nCapped = node.capped
-      nCapped.length == 32 || (node.hash sameElements previousRootHash)
-    }.map(n => ByteString(n.hash))
-    val toBeUpdated = toUpdate.filter { n =>
-      val nCapped = n.capped
-      nCapped.length == 32 || (nCapped == rootCapped)
-    }.map(node => ByteString(node.hash) -> node.encode)
-    nodeStorage.update(toBeRemoved, toBeUpdated)
-  }
-
-  private def getNextNode(extensionNode: ExtensionNode, nodeStorage: NodesKeyValueStorage)(implicit nodeDec: RLPDecoder[MptNode]): MptNode =
-    extensionNode.next match {
-      case node@(_:BranchNode | _:ExtensionNode| _:LeafNode) => node
-      case HashNode(hash) =>
-        val nodeId = hash.toArray[Byte]
-        MerklePatriciaTrie.getNode(nodeId, nodeStorage).getOrElse(throw new MissingNodeException(ByteString(nodeId)))
-      case NullNode => throw new MPTException("Extension node can't be null")
-    }
-
-  private def getChild(branchNode: BranchNode, pos: Int, nodeStorage: NodesKeyValueStorage)(implicit nodeDec: RLPDecoder[MptNode]): MptNode =
-    branchNode.children(pos) match {
-      case node@(_:BranchNode | _:ExtensionNode | _:LeafNode | _: NullNode.type) => node
-      case HashNode(hash) =>
-        val nodeId = hash.toArray[Byte]
-        MerklePatriciaTrie.getNode(nodeId, nodeStorage).getOrElse(throw new MissingNodeException(ByteString(nodeId)))
-    }
-
-  implicit val nodeEnc: RLPEncoder[MptNode] = {
-    case leaf: LeafNode => RLPList(HexPrefix.encode(nibbles = leaf.key.toArray[Byte], isLeaf = true), leaf.value)
-    case extension: ExtensionNode =>
-      RLPList(HexPrefix.encode(nibbles = extension.sharedKey.toArray[Byte], isLeaf = false), this.nodeEnc.encode(extension.next))
-    case branch: BranchNode =>
-      val encoded = new Array[RLPEncodeable](MerklePatriciaTrie.ListSize)
-      var i = 0
-      while (i < BranchNode.numberOfChildren) {
-        val nodeEncoded = this.nodeEnc.encode(branch.children(i))
-        encoded(i) = nodeEncoded
-        i = i + 1
+  private def getFromHash(nodeId: Array[Byte], source: NodesKeyValueStorage): MptNode =
```

Comment:
There are two reasons:
1. If the node requested by mpt trie is not in db ( or some cache), it is means that our trie is inconsistent state and all further operations on it do not make much sense. If we would return `None`  upstream code would not have way of knowing if value was not present in db or the trie was inconsistent.
2. We have special mechanism in regular sync which catches those exceptions, and download nodes on demand from network. (we have it because fastSync or regulars sync is not perfect and sometimes we do not have all nodes needed)